server/lomas_server/utils/collection_models.py

```diff
@@ -88,11 +89,156 @@ class DatasetsCollection(BaseModel):
+    # TODO create validator and test these, see issue #323
+    max_partition_length: Optional[Annotated[int, Field(gt=0)]] = None
+    max_influenced_partitions: Optional[Annotated[int, Field(gt=0)]] = None
+    max_partition_contributions: Optional[Annotated[int, Field(gt=0)]] = None
+
+
+class StrMetadata(ColumnMetadata):
+    """Model for string metadata"""
+
+    type: Literal["string"]
+
+
+class CategoricalColumnMetadata(ColumnMetadata):
+    """Model for categorical column metadata"""
+
+    @model_validator(mode="after")
+    def validate_categories(self):
+        """Makes sure number of categories matches cardinality."""
+        if len(self.categories) != self.cardinality:
+            raise ValueError(
+                "Number of categories should be equal to cardinality."
+            )
+        return self
+
+
+class StrCategoricalMetadata(CategoricalColumnMetadata):
+    """Model for categorical string metadata"""
+
+    type: Literal["string"]
+    cardinality: int
+    categories: List[str]
+
+
+class Precision(IntEnum):
+    """Precision of integer and float data"""
+
+    SINGLE = 32
+    DOUBLE = 64
+
+
+class BoundedColumnMetadata(ColumnMetadata):
+    """Model for columns with bounded data"""
+
+    @model_validator(mode="after")
+    def validate_bounds(self):
+        """Validates column bounds."""
+        if (
+            self.lower is not None
+            and self.upper is not None
+            and self.lower > self.upper
+        ):
+            raise ValueError("Lower bound cannot be larger than upper bound.")
+
+        return self
+
+
+class IntMetadata(BoundedColumnMetadata):
+    """Model for integer column metadata"""
+
+    type: Literal["int"]
+    precision: Precision
+    lower: int
+    upper: int
+
+
+class IntCategoricalMetadata(CategoricalColumnMetadata):
+    """Model for integer categorical column metadata"""
+
+    type: Literal["int"]
+    precision: Precision
+    cardinality: int
+    categories: List[int]
+
+
+class FloatMetadata(BoundedColumnMetadata):
+    """Model for float column metadata"""
+
+    type: Literal["float"]
+    precision: Precision
+    lower: float
+    upper: float
+
+
+class BooleanMetadata(ColumnMetadata):
+    """Model for boolean column metadata"""
+
+    type: Literal["boolean"]
+
+
+class DatetimeMetadata(BoundedColumnMetadata):
+    """Model for datetime column metadata"""
+
+    type: Literal["datetime"]  # TODO make these constants, see issue #268
+    lower: datetime
+    upper: datetime
+
+
+def get_column_metadata_discriminator(v: Any) -> str:
+    """Discriminator function for determining the type of column metadata.
+
+    Args:
+        v (Any): The unparsed column metadata (either dict or class object)
+
+    Raises:
+        ValueError: If the column type cannot be found.
+
+    Returns:
+        str: The metadata string type.
+    """
+    if isinstance(v, dict):
+        col_type = v.get("type")
+    else:
+        col_type = getattr(v, "type")
```

Comment:
when would this happen ? it is always a dict in input, no ?

Reply:
Not necessarily, it can also be a dataclass

Reply:
We never use it but that's how they implement the validators in their examples

Reply:
but if it never happens in our code maybe just `col_type = v.get("type")`. I don't see when it could be a data class ?

Reply:
The problem is that the validator functions should accept anything Metadata.model_validate receives as input. So I think if we reduce the input type of the validator function to be a dict, we are not compliant with what pydantic could potentially forward to the validator (type-wise, even if it never happens in our codebase for now).

Reply:
Actually another use case is maybe if you specify the type of a Model. E.g. you have a ColumnMetadata model that you want to "transform" into a more specific IntMetadata model.

Reply:
ok